include/crm/common/output.h

```diff
@@ -0,0 +1,394 @@
+     *         was called.  false otherwise.
+     */
+    bool (*message) (pcmk__output_t *out, const char *message_id, ...);
+
+    /*!
+     * \brief Format the output of executed command.
+     *
+     * \param out         The output functions structure.
+     * \param exit_status The exit value of the executed command.
+     * \param src         The source of this output - stdout or stderr.
+     * \param buf         The raw command output to be formatted.
+     */
+    void (*command_output) (pcmk__output_t *out, int exit_status,
+                            pcmk__cmd_output_source src, const char *buf);
+
+    /*!
+     * \brief Format an informational message that should be shown to
+     *        to the user.  Not all formatters will do this.
+     *
+     * \param out The output functions structure.
+     * \param buf The message to be printed.
+     */
+    void (*info) (pcmk__output_t *out, const char *format, ...) G_GNUC_PRINTF(2, 3);
+
+    /*!
+     * \brief Format already formatted XML.
+     *
+     * \param out  The output functions structure.
+     * \param name A name to associate with the XML.
+     * \param buf  The XML in a string.
+     */
+    void (*output_xml) (pcmk__output_t *out, const char *name, const char *buf);
+
+    /*!
+     * \brief Start a new list of items.
+     *
+     * \note For text output, this corresponds to another level of indentation.  For
+     *       XML output, this corresponds to wrapping any following output in another
+     *       layer of tags.
+     *
+     * \note If singular_noun and plural_noun are non-NULL, calling end_list will
+     *       result in a summary being added.
+     *
+     * \param out           The output functions structure.
+     * \param header        The list header.
+     * \param singular_noun When outputting the summary for a list with one item,
+     *                      the noun to use.
+     * \param plural_noun   When outputting the summary for a list with more than one
+     *                      item, the noun to use.
+     */
+    void (*begin_list) (pcmk__output_t *out, const char *header,
+                        const char *singular_noun, const char *plural_noun);
+
+    /*!
+     * \brief Format a single item in a list.
+     *
+     * \param out     The output functions structure.
+     * \param id      A name to associate with this item.
+     * \param content The item to be formatted.
+     */
+    void (*list_item) (pcmk__output_t *out, const char *id, const char *content);
+
+    /*!
+     * \brief Conclude a list.
+     *
+     * \note If begin_list was called with non-NULL for both the singular_noun
+     *       and plural_noun arguments, this function will output a summary.
+     *       Otherwise, no summary will be added.
+     *
+     * \param out The output functions structure.
+     */
+    void (*end_list) (pcmk__output_t *out);
+};
+
+bool
+pcmk__call_message(pcmk__output_t *out, const char *message_id, ...);
```

Comment:
Let's shorten this to pcmk__output() or pcmk__message() since it will be called a lot

Reply:
This area is a little confusing.  pcmk__call_message and pcmk__register_message are not really meant to be called directly.  They get assigned to function pointers in the pcmk__output_t - see pcmk__mk_xml_output for instance.  However, since they are the same across all formatters I needed to expose them somewhere under some name.  Perhaps there's a better place to put them to make this clearer.

Reply:
Ah of course, I looked right over that. Maybe just add a doxygen block mentioning that.